contrib/pkg/adm/managedns/enable.go

```diff
@@ -160,112 +171,159 @@ func (o *Options) Run(dynClient client.Client, args []string) error {
+	_, err = o.hiveClient.HiveV1().HiveConfigs().Update(hc)
 	if err != nil {
 		log.WithError(err).Fatal("error updating HiveConfig")
 	}
 	log.Info("updated HiveConfig")
 
 	// Adding manageDNS to HiveConfig triggers a new rollout of the hiveadmission pods.
-	// To know when it's safe to proceed, we will list all hiveadmission pods and ensure
-	// they all reference the new configmap.
+	// To know when it's safe to proceed, we will wait for the HiveConfig to reflect
+	// that it has been processed successfully and then do the equivalent of a
+	// kubectl rollout status --wait
 
-	err = waitForHiveConfigToBeProcessed(dynClient)
+	err = waitForHiveConfigToBeProcessed(o.hiveClient)
 	if err != nil {
 		log.WithError(err).Fatal("gave up waiting for HiveConfig to be processed")
 	}
 
-	log.Debug("getting new configmap with updated dns data")
-	dnsConfigMapName, err := getUpdatedManagedDNSConfig(dynClient)
-	if err != nil {
-		log.WithError(err).Fatal("failed to get new configmap name")
-	}
-
-	if err := waitForHiveAdmissionPods(dynClient, dnsConfigMapName); err != nil {
+	if err := waitForHiveAdmissionPods(o.dynamicClient); err != nil {
 		log.WithError(err).Fatal("hive admission pods never became available")
 	}
 
 	log.Info("Hive is now ready to create clusters with manageDNS=true")
 	return nil
 }
 
-func waitForHiveAdmissionPods(dynClient client.Client, dnsConfigMapName string) error {
+func waitForHiveAdmissionPods(dynClient dynamic.Interface) error {
+	resourceName := "deployments"
+	gvr := appsv1.SchemeGroupVersion.WithResource(resourceName)
+
 	log.Info("waiting for new hiveadmission pods to deploy")
-	selector := map[string]string{"hiveadmission": "true"}
-	for i := 0; i < 24; i++ {
-		hiveAdmissionPods := &corev1.PodList{}
-		err := dynClient.List(context.Background(), hiveAdmissionPods, client.MatchingLabels(selector))
+
+	unstDeployment, err := dynClient.Resource(gvr).Namespace(constants.HiveNamespace).Get(hiveAdmissionDeployment, metav1.GetOptions{})
+	if err != nil {
+		log.WithError(err).Error("failed to get deployment object")
+		return err
+	}
+
+	deployment := &appsv1.Deployment{}
+	if err := scheme.Scheme.Convert(unstDeployment, deployment, nil); err != nil {
+		log.WithError(err).Error("failed to convert unstructed deployment to deployment")
+		return err
+	}
+
+	statusViewer := &kubectl.DeploymentStatusViewer{}
+
+	fieldSelector := fields.OneTermEqualSelector("metadata.name", hiveAdmissionDeployment).String()
+
+	lw := &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			options.FieldSelector = fieldSelector
+			return dynClient.Resource(gvr).Namespace(constants.HiveNamespace).List(options)
+
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			options.FieldSelector = fieldSelector
+			return dynClient.Resource(gvr).Namespace(constants.HiveNamespace).Watch(options)
+		},
+	}
+
+	preconditionFunc := func(store cache.Store) (bool, error) {
```

Comment:
I don't think we need this precondition for our use case. We **want** to wait for the timeout if the deployment does not exist already.

Reply:
maybe i'm missing something here. why would a deployment not exist? this is the hiveadmission pods we're talking about which should always be running on a healthy instance of hive, no?

Reply:
This hiveutil command could be run prior to the hive-operator creating the hiveadmission deployments for the first time.